### ISSUE TITLE
Parameter Count

### ISSUE THREAD

**Issue by jafioti** [coreylowman/dfdx#435]
When training large models, it's quite useful to know the size of the model being trained. In pytorch it's fairly straightforward to calculate this:
```python
params = sum(p.numel() for p in model.parameters() if (p.requires_grad or not trainable_only))
```
This is doable because we can use the `model.parameters()` to get a list of all the `Parameter`s in the model, and then sum over the trainable ones.

Currently in dfdx I made a trait for counting parameters called `CountParams`:
```rust
pub trait CountParams {
    fn param_count(&self) -> usize;
}
```
which gets implemented on all tensors like so:
```rust
impl <
    S: Shape,
    E: Dtype + Unit,
    T,
    D: DeviceStorage + ZerosTensor<E> + CopySlice<E>
> CountParams for Tensor<S, E, D, T> {
    fn param_count(&self) -> usize {
        self.shape().num_elements()
    }
}
```

Of course the straightforward way to use this is to implement it on all modules, and then just call `.param_count()` on your model. I don't really think this is a great way to do it though. It seems for every new piece of functionality, we need to constantly implement new traits for each and every module, and as the number of modules grows, this becomes harder and harder to maintain.

One alternative might be to have a state-dict style structure like pytorch has. Something like:
```rust
trait ToStateDict {
    fn state_dict(&self) -> HashMap<String, &Tensor>;
}
```
which would allow modules to specify their storage layout. Once implemented, this could give us CountParams, as well as SaveToNpz for free, because we now have the data and the paths for each param.If the state_dict returned `HashMap<String, &mut Tensor>` instead, it would also give us LoadFromNpz for free as well, because we have the paths as well as the tensors to load into.

Not sure this is the best way, but I don't think constantly implementing new traits will be sustainable long-term.

**Comment by coreylowman** [comment]
Yeah this will require figuring out how to have a tensor struct with no generics on it. This could also help the #400 issue, which right now doesn't work because we have to store heterogeneous collection of tensors as `Box<dyn Any>`.

**Comment by coreylowman** [comment]
I think this would be huge if we could figure this out. Having a collection of tensors with different shapes/dtypes/devices where you can still actually do something with the tensors would be amazing

**Comment by jafioti** [comment]
@coreylowman Would it be possible to have certain traits on it, like `Box<dyn NumElements>` which is implemented on `Tensor` with all generics? I understand AddAssign didn't work here, but perhaps others would?

**Comment by coreylowman** [comment]
I think returning hashmap/iterator/vec of `TensorView<'a, E, D>`/`TensorViewMut<'a, E, D>` could work, but it'd require changes to `DeviceStorage` to allow runtime known shapes I think

```rust
trait DeviceStorage {
    ...
    type View<'a>;
}

struct TensorView<'a, D: DeviceStorage> {
    id: &'a UniqueId,
    view: D::View<'a>,
    device: &'a D,
}

impl Tensor<S, E, D, T> {
    pub fn view(&self) -> TensorView<'_, D> {

    }

   pub fn view_mut(&mut self) -> TensorViewMut<'_, D> {

}
}
```

**Comment by nkoppel** [comment]
Another potential idea is to have a trait named VisitTensors that takes one array of mutable references and an array of immutable references to a Module, using [array.map](https://doc.rust-lang.org/std/primitive.array.html#method.map) to recurse through the Module's fields. However, this doesn't seem to be possible.
```rust

trait VisitTensorsFn<const N: usize, const M: usize, E: Dtype, D: DeviceStorage> {
    fn call<S: Shape>(
        &mut self,
        tensors: [&Tensor<S, E, D>; N],
        tensors_mut: [&mut Tensor<S, E, D>; M],
    ) -> Result<(), D::Err>;
}

trait VisitTensors<const N: usize, const M: usize, E: Dtype, D: DeviceStorage> {
    fn visit<F: VisitTensorsFn<N, M, E, D>>(
        selfs: [&Self; N],
        selfs_mut: [&mut Self; M],
        func: &mut F,
    ) -> Result<(), D::Err>;
}

impl<
        const N: usize,
        const M: usize,
        const I: usize,
        const O: usize,
        E: Dtype,
        D: DeviceStorage,
    > VisitTensors<N, M, E, D> for crate::nn::linear::Linear<I, O, E, D>
{
    fn visit<F: VisitTensorsFn<N, M, E, D>>(
        selfs: [&Self; N],
        selfs_mut: [&mut Self; M],
        func: &mut F,
    ) -> Result<(), D::Err> {
        func.call(selfs.map(|m| &m.weight), selfs_mut.map(|m| &mut m.weight))?;
        func.call(selfs.map(|m| &m.bias), selfs_mut.map(|m| &mut m.bias))
    }
}
```
The above code fails because `array.map` takes ownership of its argument, and this consumes `selfs_mut` because mutable references cannot be copied. This code, however would be safe if selfs_mut were not consumed by map if the mutable references to the weights was dropped, which could be enforced by taking a mutable reference to the array of weight, so perhaps there is some workaround. Regardless, if we get this to work, it should be fairly simple for end users to implement and should be able to implement most of the functionality we would want from iterating through tensor references.

**Comment by coreylowman** [comment]
Oh! Visitor api!

```rust
use dfdx::prelude::*;

trait TensorVisitor {
    fn visit_mut<S: Shape, E: Dtype, D: DeviceStorage>(&mut self, tensor: &mut Tensor<S, E, D>);
}

trait TensorCollection {
    fn iter_tensors_mut<V: TensorVisitor>(&mut self, visitor: &mut V);
}

impl<const I: usize, const O: usize, E: Dtype, D: DeviceStorage> TensorCollection
    for dfdx::nn::modules::Linear<I, O, E, D>
{
    fn iter_tensors_mut<V: TensorVisitor>(&mut self, visitor: &mut V) {
        visitor.visit_mut(&mut self.weight);
        visitor.visit_mut(&mut self.bias);
    }
}

struct CountParams(usize);
impl TensorVisitor for CountParams {
    fn visit_mut<S: Shape, E: Dtype, D: DeviceStorage>(&mut self, tensor: &mut Tensor<S, E, D>) {
        self.0 += tensor.shape().num_elements();
    }
}

struct NPZSerialize;
impl TensorVisitor for NPZSerialize {
    fn visit_mut<S: Shape, E: Dtype, D: DeviceStorage>(&mut self, tensor: &mut Tensor<S, E, D>) {
        todo!()
    }
}
```

**Comment by coreylowman** [comment]
I think we'd likely want to include a notion of naming with `visit_tensors` to support the npz serialization, and then it'd be great to figure out how we can do GradientUpdate with this approach.

It seems possible to also handle ResetParams with tensor visitor, but we'd have to have some sort of registry of how to reset parameters? And ideally that would be aligned with BuildModule as well.

**Comment by coreylowman** [comment]
A slight complication for gradient update using some sort of visitor/iterator pattern: BatchNorm2D 😩. We'd want to include the running_mean/running_std for things like npz serialization, but not include them for things like GradientUpdate.

**Comment by nkoppel** [comment]
One way we could deal with this is by adding a method to TensorVisitor that allows modules to configure the visitor as it runs:
```rust
enum VisitorOption {
    ...
}
trait TensorVisitor {
    fn set_option(&mut self, option: VisitorOption) {}

    ...
}
```
I've also considered cloning TensorVisitors with new options, but some visitors will need to hold mutable references, so it will not be possible to clone them. Unfortunately, this does mean that we have to remember to set each option back to its default before the end of the call to visit.

**Comment by coreylowman** [comment]
Yeah, still haven't been able to figure out how to use a visitor approach to do EMA either. I'm stuck on each module having to implement "zipped" visitors as well, which is a big downside (and no different than just implementing some EMA trait directly).

I do feel like there's some clever trick we could apply to make it work though...
